### PR TITLE
feat: Reworked interpolation_search.cpp

### PR DESCRIPTION
**Pull request by Lazeeez**
#### Description of Change

The original `interpolation_search.cpp` was pretty blatant with little to no documentation, did some rework on it and added a testcase to check the algorithm.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Did some rework on the original `interpolation_search.cpp`

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1855"></a>

### CODE REVIEW

**Comment by Lazeeez** [comment]
@Panquesito7 Can you please review this PR, thanks!

File: search/interpolation_search.cpp [hunk 1]
```diff
@@ -1,61 +1,121 @@
...
-            low = mid + 1;
-        else
-            return mid;
-    }
 
-    if (arr[low] == value)
-        return low;
+/******************************************************************************
+ * @file
+ * @brief [interpolation search
```
Comment (Panquesito7) [inline search/interpolation_search.cpp:4]:
```suggestion
 * @brief [Interpolation search
```

File: search/interpolation_search.cpp [hunk 2]
```diff
@@ -1,61 +1,121 @@
...
+
+/******************************************************************************
+ * @namespace search
+ * @brief Searching algorithms
+ *******************************************************************************/
+namespace search {
+
+/******************************************************************************
+ * @namespace interpolation_search
+ * @brief interpolation search searching algorihm
```
Comment (Panquesito7) [inline search/interpolation_search.cpp:45]:
```suggestion
 * @brief Functions for the [Interpolation Search](https://en.wikipedia.org/wiki/interpolation_search) algorithm implementation
```

File: search/interpolation_search.cpp [hunk 3]
```diff
@@ -1,61 +1,121 @@
...
+ *******************************************************************************/
+namespace interpolation_search {
 
-    std::cout << "Enter the size of array(less than 100) : ";
-    std::cin >> n;
+/******************************************************************************
+ * @brief The main function which implements binary search
```
Comment (Panquesito7) [inline search/interpolation_search.cpp:50]:
```suggestion
 * @brief The main function which implements interpolation search
```

File: search/interpolation_search.cpp [hunk 4]
```diff
@@ -1,61 +1,121 @@
...
-    else
-        std::cout << "The value is at the position " << re << std::endl;
+}  // namespace search
 
-    delete[] array;
+/*******************************************************************************
+ * @brief Self-test implementation
+ * @returns void
+ *******************************************************************************/
+void tests() {
```
Comment (Panquesito7) [inline search/interpolation_search.cpp:98]:
```suggestion
static void tests() {
```

File: search/interpolation_search.cpp [hunk 5]
```diff
@@ -1,61 +1,121 @@
...
+    assert(derived_answer == expected_answer);
+    std::cout << "Passed!";
+}
+
+/*******************************************************************************
+ * @brief Main function
+ * @returns 0 on exit
+ *******************************************************************************/
+int main() {
+    tests();  // Condition of target found
```
Comment (Panquesito7) [inline search/interpolation_search.cpp:119]:
```suggestion
    tests();  // run self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 🚀
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/Lazeeez/C-Plus-Plus/actions

**Comment by Lazeeez** [comment]
> Great work! 🚀 Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/Lazeeez/C-Plus-Plus/actions

The Github actions are already enabled for this entire repository.

**Comment by Lazeeez** [comment]
@Panquesito7 check now.

File: search/interpolation_search.cpp [hunk 6]
```diff
@@ -1,61 +1,123 @@
...
-        std::cout << "The value is at the position " << re << std::endl;
+}  // namespace search
 
-    delete[] array;
+/*******************************************************************************
+ * @brief Self-test implementation
+ * @returns void
+ *******************************************************************************/
+static void tests() {
+    // testcase
```
Comment (Panquesito7) [inline search/interpolation_search.cpp:101]:
```suggestion
    // 1st test
```

File: search/interpolation_search.cpp [hunk 7]
```diff
@@ -1,61 +1,123 @@
...
+    // array = [10, 12, 13, 16, 18, 19, 20, 21, 1, 2, 3, 4, 22, 23, 24, 33, 35,
+    // 42, 47] , Value = 33 should return 15
+    std::vector<uint64_t> arr = {{10, 12, 13, 16, 18, 19, 20, 21, 1, 2, 3, 4,
+                                  22, 23, 24, 33, 35, 42, 47}};
+    sort(arr.begin(), arr.end());
+    uint64_t number = 33;  // Element to be searched
+    uint64_t expected_answer = 15;
+    uint64_t derived_answer =
+        search::interpolation_search::interpolationSearch(arr, number);
+    std::cout << "Testcase : ";
```
Comment (Panquesito7) [inline search/interpolation_search.cpp:111]:
```suggestion
    std::cout << "1st test: ";
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: search/interpolation_search.cpp (hunk 6 above)
Comment (Panquesito7) [inline search/interpolation_search.cpp:101]:
```suggestion
    // 1st test
```

File: search/interpolation_search.cpp (hunk 7 above)
Comment (Panquesito7) [inline search/interpolation_search.cpp:111]:
```suggestion
    std::cout << "1st test: ";
```

File: search/interpolation_search.cpp [hunk 8]
```diff
@@ -1,61 +1,123 @@
...
+    assert(derived_answer == expected_answer);
+    std::cout << "Passed!";
```
Comment (Panquesito7) [inline search/interpolation_search.cpp:113]:
```suggestion
    std::cout << "Passed!\n";
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: search/interpolation_search.cpp (hunk 8 above)
Comment (Panquesito7) [inline search/interpolation_search.cpp:113]: Let's add a new line here. It'll look better in the terminal. 😃
```suggestion
    std::cout << "Passed!\n";
```

File: search/interpolation_search.cpp (hunk 7 above)
Comment (Panquesito7) [inline search/interpolation_search.cpp:111]:
```suggestion
    std::cout << "Testcase: ";
```

**Review by Panquesito7 — APPROVED** [review]
LGTM 🚀 Thank you for your contribution! 😄👍

**Review by Qannaf — APPROVED** [review]

**Review by ayaankhan98 — APPROVED** [review]
LGTM

**Review by Panquesito7 — APPROVED** [review]

**Pull request merged**